code/game/gamemodes/miniantags/pulsedemon/pulsedemon.dm

```diff
@@ -0,0 +1,713 @@
+		// don't sanitize again
+		current_robot.say(message, null, FALSE, ignore_speech_problems, ignore_atmospherics, ignore_languages)
+		return TRUE
+
+	var/message_mode = parse_message_mode(message, "headset")
+
+	if(message_mode)
+		if(message_mode == "headset")
+			message = copytext(message, 2)
+		else
+			message = copytext(message, 3)
+
+	message = trim_left(message)
+
+	var/list/message_pieces = list()
+	if(ignore_languages)
+		message_pieces = message_to_multilingual(message)
+	else
+		message_pieces = parse_languages(message)
+
+	// hivemind languages
+	if(istype(message_pieces, /datum/multilingual_say_piece))
+		var/datum/multilingual_say_piece/S = message_pieces
+		S.speaking.broadcast(src, S.message)
+		return TRUE
+
+	if(!LAZYLEN(message_pieces))
+		. = FALSE
+		CRASH("Message failed to generate pieces. [message] - [json_encode(message_pieces)]")
+
+	create_log(SAY_LOG, "[message_mode ? "([message_mode])" : ""] '[message]'")
+
+	playsound(get_turf(src), pick(speech_sounds), 50, TRUE)
+	if(istype(loc, /obj/item/radio))
+		var/obj/item/radio/R = loc
+		name = gen_speech_name()
+		R.talk_into(src, message_pieces, message_mode, verbage = verb)
+		name = real_name
+		return TRUE
+	else if(istype(loc, /obj/machinery/hologram/holopad))
+		var/obj/machinery/hologram/holopad/H = loc
+		var/turf/T = get_turf(H)
+		name = "[H]"
+		for(var/mob/M in hearers(T.loc) + src)
+			M.hear_say(message_pieces, verb, FALSE, src)
+		name = real_name
+		return TRUE
+
+	emote("me", message = "[pick(emote_hear)]")
+	return TRUE
+
+/mob/living/simple_animal/pulse_demon/update_runechat_msg_location()
+	if(istype(loc, /obj/machinery/hologram/holopad))
+		runechat_msg_location = loc.UID()
+	else
+		return ..()
+
+/mob/living/simple_animal/pulse_demon/has_internal_radio_channel_access(mob/user, list/req_one_accesses)
+	var/list/access = get_all_accesses()
+	return has_access(list(), req_one_accesses, access)
+
+/mob/living/simple_animal/pulse_demon/proc/try_hijack_apc(obj/machinery/power/apc/A, remote = FALSE)
+	// one APC per pulse demon, one pulse demon per APC, no duplicate APCs
+	if(check_valid_apc(A, loc) && !(A in hijacked_apcs) && !apc_being_hijacked && !A.being_hijacked)
+		to_chat(src, "<span class='notice'>You are now attempting to hijack [A], this will take approximately [hijack_time / 10] seconds.</span>")
+		if(pb_helper.start(src, A, hijack_time, TRUE, \
+		  CALLBACK(src, PROC_REF(check_valid_apc), A), \
+		  CALLBACK(src, PROC_REF(finish_hijack_apc), A, remote), \
+		  CALLBACK(src, PROC_REF(fail_hijack)), \
+		  CALLBACK(src, PROC_REF(cleanup_hijack_apc), A)))
+			apc_being_hijacked = A
+			A.being_hijacked = TRUE
+			A.update_icon()
+			return TRUE
+		else
+			to_chat(src, "<span class='warning'>You are already performing an action!</span>")
+	return FALSE
+
+/mob/living/simple_animal/pulse_demon/proc/check_valid_apc(obj/machinery/power/apc/A, atom/startloc)
+	return is_valid_apc(A)
```

Comment:
is this abstraction unneeded?
why can't we just directly call `is_valid_apc(A)` instead of needing to call `check_valid_apc(A)`, the second parameter isn't even being used in this case